components/prism-liquid.js

```diff
@@ -1,12 +1,61 @@
 Prism.languages.liquid = {
-	'keyword': /\b(?:comment|endcomment|if|elsif|else|endif|unless|endunless|for|endfor|case|endcase|when|in|break|assign|continue|limit|offset|range|reversed|raw|endraw|capture|endcapture|tablerow|endtablerow)\b/,
-	'number': /\b0b[01]+\b|\b0x(?:\.[\da-fp-]+|[\da-f]+(?:\.[\da-fp-]+)?)\b|(?:\b\d+(?:\.\d*)?|\B\.\d+)(?:e[+-]?\d+)?[df]?/i,
-	'operator': {
-		pattern: /(^|[^.])(?:\+[+=]?|-[-=]?|!=?|<<?=?|>>?>?=?|==?|&[&=]?|\|[|=]?|\*=?|\/=?|%=?|\^=?|[?:~])/m,
+	'comment': {
+		pattern: /(^\{%\s*comment\s*%\})[\s\S]+(?=\{%\s*endcomment\s*%\}$)/,
 		lookbehind: true
 	},
-	'function': {
-		pattern: /(^|[\s;|&])(?:append|prepend|capitalize|cycle|cols|increment|decrement|abs|at_least|at_most|ceil|compact|concat|date|default|divided_by|downcase|escape|escape_once|first|floor|join|last|lstrip|map|minus|modulo|newline_to_br|plus|remove|remove_first|replace|replace_first|reverse|round|rstrip|size|slice|sort|sort_natural|split|strip|strip_html|strip_newlines|times|truncate|truncatewords|uniq|upcase|url_decode|url_encode|include|paginate)(?=$|[\s;|&])/,
-		lookbehind: true
-	}
+	'delimiter': {
+		pattern: /^\{(?:\{\{|[%\{])-?|-?(?:\}\}|[%\}])\}$/,
+		alias: 'punctuation'
+	},
+	'string': {
+		pattern: /"[^"]*"|'[^']*'/,
+		greedy: true
+	},
+	'keyword': /\b(?:as|assign|break|continue|cycle|decrement|echo|else|elsif|(?:end)?(?:capture|case|comment|for|form|if|paginate|style|raw|tablerow|unless)|in|include|increment|limit|liquid|offset|range|render|reversed|section|when|with)\b/,
+	'function': [
+		{
+			pattern: /(\|\s*)\w+/,
+			lookbehind: true,
+			alias: 'filter'
+		},
+		{
+			// array functions
+			pattern: /(\.\s*)(?:first|last|size)/,
+			lookbehind: true
+		}
+	],
+	'boolean': /\b(?:true|false|nil)\b/,
+	'range': {
+		pattern: /\.\./,
+		alias: 'operator'
+	},
+	// https://github.com/Shopify/liquid/blob/698f5e0d967423e013f6169d9111bd969bd78337/lib/liquid/lexer.rb#L21
+	'number': /\b\d+(?:\.\d+)?\b/,
+	'operator': /[!=]=|<>|[<>]=?|[|?:=-]|\b(?:and|or|contains(?=\s))\b/,
+	'punctuation': /[.,\[\]()]/
 };
+
+Prism.hooks.add('before-tokenize', function (env) {
+	var liquidPattern = /\{%\s*comment\s*%\}[\s\S]*?\{%\s*endcomment\s*%\}|\{(?:%[\s\S]*?%|\{\{[\s\S]*?\}\}|\{[\s\S]*?\})\}/g;
+	var insideRaw = false;
+
+	Prism.languages['markup-templating'].buildPlaceholders(env, 'liquid', liquidPattern, function (match) {
```

Comment:
If I were to wish for something it would be that the error message would tell you that you need to have `prism-markup-templating` loaded, something like this

```suggestion
	if (!Prism.languages['markup-templating']) {
		throw new Error('You need to load prismjs/components/prism-markup-templating to use the Liquid lang'); 
	}

	Prism.languages['markup-templating'].buildPlaceholders(env, 'liquid', liquidPattern, function (match) {
```

Reply:
Languages don't load their own dependencies. It's not their responsibility to ensure the correct load order. If Prism components aren't loaded correctly, they are allowed to fail for any reason.

I get that good error messages are nice in these situations but we have a hundred languages with dependencies like this. If all of them checked their dependencies and threw an error like this, then Prism would carry a lot of dead weight for a lot of users (namely: all users that correctly use Prism, so hopefully all users).

We want Prism to be as light-weight as possible and this just doesn't make the cut.

Reply:
hm. Interesting. Is the assumption that everyone is loading _every_ language/component?

Reply:
No, only the ones wanted + their dependencies. Why?

Reply:
> No, only the ones wanted + their dependencies. Why?

In that case, I don't think this statement is true:

> If all of them checked their dependencies and threw an error like this, then Prism would carry a lot of dead weight for a lot of users

It would certainly add a few bytes here and there. However, I think the improved developer experience is well worth the trade.

Reply:
> However, I think the improved developer experience is well worth the trade.

I disagree. Instead of telling you "you should have loaded these files", the components should just load their dependencies themselves. This is how it will be done in v2.0. So we won't have to worry about this for much longer.

Until then, I would like to carry on as usual. I mean, it seems to have worked so far ;)

Reply:
Haha that’s fair. I didn’t want to suggest figuring out dependency injection but glad that’s in the works!